### PR TITLE
feat: spring docs pubsub #7722

### PR DESCRIPTION
**Pull request by franciscolopezsancho**
References https://github.com/lightbend/kalix/issues/7722

### CODE REVIEW

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [inline docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:100]: https://github.com/lightbend/kalix/issues/7799

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (efgpinto) [inline docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:15]: I think all this section is a good candidate to have in a partial shared between java and spring docs.. It's very self-contained and can't see it change a lot.

small independent note: since we link to google pub/sub, should we add a link to kafka too on the first paragraph?

File: samples/spring-eventsourced-counter/src/main/java/com/example/actions/CounterJournalToTopicAction.java
Comment (efgpinto) [inline samples/spring-eventsourced-counter/src/main/java/com/example/actions/CounterJournalToTopicAction.java:21]: Maybe a easier way here is to use a specific tag for each import and then use that tag on docs.. when importing the file, you can pass in several tags... e.g. `CounterJournalToTopicAction.java[tag=importX,imporY,sub-ESE-pub-topic-action]`

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (efgpinto) [inline docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:32]:
```suggestion
The events from the entity are delivered to the Action and what to do with those is up for implementation. For this simplified example you are just logging it, but it could also do a forward to some other component or external service.
```

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (efgpinto) [inline docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:42]:
```suggestion
You can subscribe an Action to events from a Value Entity. To receive messages from the entity, annotate a service method `@Subscribe.ValueEntiy` and specify the class of the entity. The same applies as for subscribing to an Event Sourced Entity.
```

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (efgpinto) [inline docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:47]:
```suggestion
To receive messages from a Google Cloud Pub/Sub or Apache Kafka topic, annotate the service method `@Subscribe.Topic` and specify the topic name.
```

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (efgpinto) [inline docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:61]: Make use of the "NOTE" special section

```suggestion
NOTE: by default, Kalix assumes the message in the topic was serialized as JSON and as such deserialize it into a `CounterEvent`. You can find more about serialization in Kalix here xref:spring:serialization.adoc[].
```

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (efgpinto) [inline docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:67]: I'm actually not sure we support this yet for the spring-sdk.. do we?

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (efgpinto) [inline docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:73]: This link `xref:actions-as-controller.adoc[Actions as Controller]` will probably need to be spring specific?

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (efgpinto) [inline docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:91]: Maybe use `IMPORTANT:` here?

```suggestion
IMPORTANT: the messages stored in the topic are serialized as JSON. You cannot change this serialization.
```

**Review by efgpinto — COMMENTED** [review]
I think I might not be the best person to review because I haven't played much with publish/subscribe so I might be missing some details.

But looking good... Left some feedback. 
I think you will need to add some broker configs so the sample can publish stuff in CI.

File: samples/spring-eventsourced-counter/src/main/java/com/example/actions/CounterJournalToTopicAction.java
Comment (franciscolopezsancho) [reply samples/spring-eventsourced-counter/src/main/java/com/example/actions/CounterJournalToTopicAction.java:21]: I didn't know that :) awesome! will do

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:67]: You are right, not yet. This was an idea I talked about with Renato but is not yet implemented.

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:91]: Yeah, this is a good point my only concern when writing this and the 'Note that' above - that you suggested adding a NOTE: - is that it will lead the code with no leading text and looks somehow abrupt. But I agree it's important the reader notes it.

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:15]: It makes sense. Do you know if we other partial common to Java and Spring already?

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:61]: I used it already I was feeling I would be abusing it.

Comment (efgpinto) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:61]: I haven't actually rendered it, so you certainly have a better perception..

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:73]: I'll check it out

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:67]: removed

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:73]: fixed

File: samples/spring-eventsourced-counter/src/main/java/com/example/actions/CounterJournalToTopicAction.java
Comment (franciscolopezsancho) [reply samples/spring-eventsourced-counter/src/main/java/com/example/actions/CounterJournalToTopicAction.java:21]: I tried and didn't seem to work. Do you have a sample of it? I'm looking at the Antora docs and can't find the proper info. There is some entry about 'Tags' but that's not what we are using here I think.

Comment (efgpinto) [reply samples/spring-eventsourced-counter/src/main/java/com/example/actions/CounterJournalToTopicAction.java:21]: My bad, correct form is `tags` and using `;` as separator.. so: CounterJournalToTopicAction.java[tags=importX;importY;sub-ESE-pub-topic-action]

See example in views.adoc#L46 from java [here](https://github.com/lightbend/kalix-jvm-sdk/blob/main/docs/src/modules/java/pages/views.adoc#L46)

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:15]: I'm having second thoughts. I prefer if we leave this optimization after we have the docs in better shape.

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:61]: I used your suggested below "IMPORTANT".

File: samples/spring-eventsourced-counter/src/main/java/com/example/actions/CounterJournalToTopicAction.java
Comment (franciscolopezsancho) [reply samples/spring-eventsourced-counter/src/main/java/com/example/actions/CounterJournalToTopicAction.java:21]: I had to divide it in two classes so I don't need it for now but I keep it in mind. It's very neat 👍

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:15]: :) I did add the link in a paragraph below. I move it from there to here.

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (efgpinto) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:73]: I think I was wrong about this one.. I just tested and with this link, by default it goes to the page under the same module...

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (efgpinto) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:15]: > Do you know if we have other partials common to Java and Spring already?

Yeah, I did it for the intro in eventsourced entities and I think I will do the same for timers. But we can evaluate later if we think it's worth or not.

**Comment by franciscolopezsancho** [comment]
I think we won't be able to pass CI until we provide Pub/Sub capabilities to the test kit. Here I added an [issue](https://github.com/lightbend/kalix/issues/7811) for it.

**Review by octonato — COMMENTED** [review]
Left some comments, I think for our sake, we should keep the structure as close as possible to the gRPC counter-part. 

If we believe the gRPC one is unclear or lacking, then we should create a ticket to improve it later.

**Comment by octonato** [comment]
> I think we won't be able to pass CI until we provide Pub/Sub capabilities to the test kit. Here I added an [issue](https://github.com/lightbend/kalix/issues/7811) for it.

I didn't see any an integration test that would required Pub/Sub support on the test kit. Is that something you were working on when you wrote this comment and afterwards postponed it? 

In any case, the gRPC one doesn't have it either, so I wouldn't block this PR because of lack of testkit support for it. Let's just move on, create ticket and tackle it later.

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:37]: I think is important to explain that to the user even if we return `Confirmed()`. Otherwise, the user will have the wrong model of how this works. It is very useful for me to have this model in mind and I remember when I found out that it was an aha moment. 
I'm happy to introduce a record  like`Confirmed` if you think is better. Please, let me know.

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:19]: ok. will do.

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:40]: yup

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:90]: Agree, although I think this is a general idea. It may be useful for the reader to abstract upon the first example. FYI This is how is it in the original docs. Would you like to remove it?

**Comment by franciscolopezsancho** [comment]
Is not that there is a specific integration test for Pub/Sub. Here is the [error](https://app.circleci.com/pipelines/github/lightbend/kalix-jvm-sdk/4658/workflows/b0925bdc-2c01-4788-b9c0-92cf5f5990da/jobs/76464)

It's a 500 with little explanation but what I think it's happening is that because there is an Action Subscribing/Publish to a topic when the proxy receives this info tries to connect with the underlying Pub/Sub system and because there is not it fails miserably.

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (octonato) [inline docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:33]: Let's just say here that we return `Confirmed`

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (octonato) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:42]: Good to explain it. Good idea. 

We can convert this NOTE into an explanation about the meaning of returning `effects().reply(Confirmed.instance)` or returning `effects().error("something went wrong")` or even throwing an exception.

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [inline docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:47]:
```suggestion
In the following example the events from the topic are delivered to the Action and logged.
```

File: sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/Subscribe.java
Comment (efgpinto) [inline sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/Subscribe.java:77]: nice

**Review by efgpinto — APPROVED** [review]
LGTM, just tiny grammar appoiments.

File: docs/src/modules/spring/pages/actions-publishing-subscribing.adoc
Comment (franciscolopezsancho) [reply docs/src/modules/spring/pages/actions-publishing-subscribing.adoc:13]: good catch 👍

File: samples/spring-eventsourced-counter/src/main/java/com/example/actions/Confirmed.java
Comment (franciscolopezsancho) [reply samples/spring-eventsourced-counter/src/main/java/com/example/actions/Confirmed.java:13]: yeah, it does look weird. I'm checking the source file and it's only one tab. What can this be? I'll find out.

File: sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/Subscribe.java
Comment (franciscolopezsancho) [reply sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/Subscribe.java:77]: :)

File: .circleci/config.yml
Comment (franciscolopezsancho) [reply .circleci/config.yml]: good catch! fixed now

File: samples/spring-eventsourced-counter/src/main/java/com/example/actions/Confirmed.java
Comment (franciscolopezsancho) [reply samples/spring-eventsourced-counter/src/main/java/com/example/actions/Confirmed.java:13]: I was too confident saying I was going to find out. I did not :) I think it may be GitHub. I'll double-check after the merge to the main that it is one tab.

**Pull request merged**
